core/imds/src/main/java/software/amazon/awssdk/imds/internal/DefaultEc2Metadata.java

```diff
@@ -152,78 +154,143 @@ public String toString() {
 
-            if (statusCode == HttpURLConnection.HTTP_OK && responseBody.isPresent()) {
-                abortableInputStream = responseBody.get();
-                data = IoUtils.toUtf8String(abortableInputStream);
-                metadataResponse = new MetadataResponse(data);
-            } else if (statusCode == HttpURLConnection.HTTP_NOT_FOUND) {
-                throw SdkServiceException.builder()
-                                         .message("The requested metadata at path ( " + path + " ) is not found ").build();
-            } else if (statusCode == HttpURLConnection.HTTP_OK) {
-                throw SdkClientException.builder()
-                                         .message("Response body empty with Status Code " + statusCode).build();
-            } else {
-                throw SdkClientException.builder()
-                                        .message("Instance metadata service returned unexpected status code " + statusCode)
-                                        .build();
+        for (int tries = 1 ; tries <= retryPolicy.numRetries() ; tries ++) {
+
+            try {
+                String token = getToken();
+                HttpExecuteResponse response = getDataHttpResponse(path, token);
+
+                int statusCode = response.httpResponse().statusCode();
+                Optional<AbortableInputStream> responseBody = response.responseBody();
+
+                if (statusCode == 200 && responseBody.isPresent()) {
+                    abortableInputStream = responseBody.get();
+                    String data = IoUtils.toUtf8String(abortableInputStream);
+                    metadataResponse = new MetadataResponse(data);
+                } else {
+                    doDataExceptionBehavior(statusCode, path);
+                }
+            } catch (SdkServiceException sd) {
+                throw SdkServiceException.builder().message(sd.getMessage()).cause(sd).build();
+            } catch (IOException | SdkClientException io) {
+
+                try {
+                    doRetryBehavior(tries, io);
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    throw SdkClientException.builder().message(io.getMessage()).cause(io).build();
+                }
+
+            } finally {
+                IoUtils.closeQuietly(abortableInputStream, log);
             }
-        } catch (SdkServiceException sd) {
-            throw SdkServiceException.builder().message(sd.getMessage()).cause(sd).build();
-        } catch (IOException | SdkClientException  io) {
-            // TODO Retry Logic will be added
-            log.warn("Received an IOException {0} " , io);
-        } finally {
-            IoUtils.closeQuietly(abortableInputStream, log);
         }
-
         return metadataResponse;
     }
 
+    private void doDataExceptionBehavior(int statusCode, String path) {
```

Comment:
nit: `handleErrorResponse`?

Instead of using exception as control flow, can we only throw exceptions for non-retryable cases? This way, the retryable cases will just get retried in the for loop